### ISSUE TITLE
Save Images to Server

### ISSUE THREAD

**Issue by gadget-man** [ljmerza/frigate_plate_recognizer#21]
Hi,

Is there a way to save the image sent to PlateRecognizer (showing the numberplate detected), either when a plate is detected, or optionally on every detection request (even if no plate recognised)?    There is a similar feature on https://github.com/robmarkcole/HASS-plate-recognizer but frigate_plate_recognizer seems much more efficient (and quick) so I'm not really using the HA version anymore, but really missing having a folder with all the saved plate data on it!

**Comment by gadget-man** [comment]
I've taken a fork of the repo and have been working on some tweaks to the code to allow this. Will hopefully push a PR for review in the next few days, just need to do a bit more testing.

**Comment by kyle4269** [comment]
> I've taken a fork of the repo and have been working on some tweaks to the code to allow this. Will hopefully push a PR for review in the next few days, just need to do a bit more testing.

I did check out your current fork. I like how it draws a rectangle around the license plate with the plate number.

**Comment by DrSpaldo** [comment]
This would be a nice addition. I would love a way to keep all plates, recognised or not. Even better would be an easy way to search by plate...

**Comment by gadget-man** [comment]
I've added a PR for this: https://github.com/ljmerza/frigate_plate_recognizer/pull/24

@DrSpaldo I didn't do anything to allow saving by plate. However an easy solution would be to add the recognised plate in the filename of the saved snapshot?

**Comment by DrSpaldo** [comment]
> I've added a PR for this: #24
> 
> @DrSpaldo I didn't do anything to allow saving by plate. However an easy solution would be to add the recognised plate in the filename of the saved snapshot?

Nice PR @gadget-man :)

Um, I’m not sure the best way but that wouldn’t be a bad idea?

**Comment by gadget-man** [comment]
I've tweaked the PR so that if a plate is recognised, it will add it to the file path when `save_timestamped_file: True`

**Comment by ljmerza** [comment]
added to v1.8.0

**Comment by DrSpaldo** [comment]
I am not sure if I have configured it incorrectly or what is happening. I am using Unraid, so I was previously using the config that is similar to Synology. I am getting the following errors when I start the container/stack:

```
File "/usr/src/app/./index.py", line 143
if not config['frigate'].get('save_snapshots', False)
^
SyntaxError: invalid syntax
File "/usr/src/app/./index.py", line 143
if not config['frigate'].get('save_snapshots', False)
^
SyntaxError: invalid syntax
File "/usr/src/app/./index.py", line 143
if not config['frigate'].get('save_snapshots', False)
^
SyntaxError: invalid syntax
File "/usr/src/app/./index.py", line 143
if not config['frigate'].get('save_snapshots', False)
^
SyntaxError: invalid syntax
File "/usr/src/app/./index.py", line 143
if not config['frigate'].get('save_snapshots', False)
^
SyntaxError: invalid syntax
File "/usr/src/app/./index.py", line 143
if not config['frigate'].get('save_snapshots', False)
^
SyntaxError: invalid syntax
File "/usr/src/app/./index.py", line 143
if not config['frigate'].get('save_snapshots', False)
^
SyntaxError: invalid syntax
```

Here is my stack compose file:

```
services:
  frigate_plate_recognizer:
    image: lmerza/frigate_plate_recognizer:latest
    container_name: frigate_plate_recognizer
    volumes:
      - /mnt/user/appdata/plate-recognizer/config:/usr/src/app/config
      - /mnt/user/appdata/plate-recognizer/plates:/usr/src/app/plates:rw
    restart: unless-stopped
    environment:
      - TZ=Australia/Brisbane
```

Here is my config.yaml:

frigate:
  frigate_url: http://192.168.2.20:5000
  mqtt_server: 192.168.2.20
  mqtt_auth: true
  mqtt_username: mosquitto
  mqtt_password: PASSWORD
  main_topic: frigate
  return_topic: plate_recognizer
  frigate_plus: false
  camera:
    - street3
  objects:
    - car
  min_score: .7
  save_snapshots: true
  draw_box: true
  always_save_snapshot: true
code_project:
  api_url: http://192.168.1.28:32168/v1/image/alpr
logger_level: INFO

**Comment by DrSpaldo** [comment]
@gadget-man @ljmerza Also tried running it in Portainer. Still getting the same error messages

**Comment by gadget-man** [comment]
there are quite a few changes in the finalised 1.8.0 that differ from my PR. For example, all references to `always_save_snapshot` have been removed, it no longer includes the Plate ID in the filename etc. 

I'm still running the version based on my PR. Let me try 1.8.0 and see if I can replicate your problem.

**Comment by gadget-man** [comment]
Yep, to confirm, I'm getting the same issue on 1.8.0. @ljmerza can you please check?

**Comment by kyle4269** [comment]
I think we need a : after 

'''
if not config['frigate'].get('save_snapshots', False)

Should be if I'm not mistaken 

if not config['frigate'].get('save_snapshots', False):
'''

**Comment by DrSpaldo** [comment]
> Yep, to confirm, I'm getting the same issue on 1.8.0. @ljmerza can you please check?

Glad I wasn’t going crazy!

**Comment by gadget-man** [comment]
> I think we need a : after
> 
> ```
> if not config['frigate'].get('save_snapshots', False)
> ```
> 
> If I'm not mistaken it should be
> 
> ```
> if not config['frigate'].get('save_snapshots', False):
> ```

Well spotted! Yep, that'll be it.

**Comment by ljmerza** [comment]
fixed in 1.8.1. there  were a few edges cases around users not having frigate+ or no plate recognized that needed to be handled. dont have much time so didn want o go back and forth in he PR. thanks for the work on this

**Comment by gadget-man** [comment]
Sorry, just ran a check and still not working:

`
2024/01/13 18:04:42 | stderr | UnboundLocalError: local variable 'frigate_url' referenced before assignment
-- | -- | --
2024/01/13 18:04:42 | stderr | snapshot_url = f"{frigate_url}/api/events/{frigate_event}/snapshot.jpg"
2024/01/13 18:04:42 | stderr | File "/usr/src/app/./index.py", line 215, in on_message
`

**Comment by gadget-man** [comment]
Looks like `frigate_url` is now being referenced in 215 but it's not declared until line 240.

**Comment by ljmerza** [comment]
sorry about that. forget python doesnt eval syntax errors until runtime. fixed in 1.8.2

**Comment by ljmerza** [comment]
added a bunch of tests and found some other bugs. pushed v1.8.3 hopefully this is it. need to add more tests in the future

**Comment by DrSpaldo** [comment]
Thanks for looking at it @ljmerza and doing the new builds. I am testing it now...

Do these messages mean anything bad?

```
Traceback (most recent call last):
File "/usr/src/app/./index.py", line 403, in <module>
main()
File "/usr/src/app/./index.py", line 399, in main
run_mqtt_client()
File "/usr/src/app/./index.py", line 358, in run_mqtt_client
mqtt_client.loop_forever()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1756, in loop_forever
rc = self._loop(timeout)
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1164, in _loop
rc = self.loop_read()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1556, in loop_read
rc = self._packet_read()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 2439, in _packet_read
rc = self._packet_handle()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3033, in _packet_handle
return self._handle_publish()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3327, in _handle_publish
self._handle_on_message(message)
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3570, in _handle_on_message
on_message(self, self._userdata, message)
File "/usr/src/app/./index.py", line 215, in on_message
snapshot_url = f"{frigate_url}/api/events/{frigate_event}/snapshot.jpg"
UnboundLocalError: local variable 'frigate_event' referenced before assignment
```

**Comment by DrSpaldo** [comment]
@gadget-man how are the plates being stored? are they in folders or just all dumped into the same folder? 

I get roughly about 400 per day. So it would end up being a very stacked folder if they are. My log is 7600 lines after 11 days...

**Comment by gadget-man** [comment]
> Thanks for looking at it @ljmerza and doing the new builds. I am testing it now...
> 
> I am getting the following error continually every time a plate is detected:
> 
> ```
> 2024-01-14 11:54:51,040 - __main__ - INFO - Time: 2024-01-14 11:54:51.040
> 2024-01-14 11:54:51,040 - __main__ - INFO - Python Version: 3.9.18 (main, Jan 11 2024, 11:28:41)
> 
> [GCC 12.2.0]
> 2024-01-14 11:54:51,040 - __main__ - INFO - Frigate Plate Recognizer Version: 1.8.3
> 
> 2024-01-14 11:54:51,040 - __main__ - INFO - Using CodeProject.AI API
> 2024-01-14 11:54:51,040 - __main__ - INFO - Starting MQTT client. Connecting to: 192.168.2.20
> 2024-01-14 11:54:51,041 - __main__ - INFO - MQTT Connected
> Traceback (most recent call last):
> File "/usr/src/app/./index.py", line 403, in <module>
> main()
> File "/usr/src/app/./index.py", line 399, in main
> run_mqtt_client()
> File "/usr/src/app/./index.py", line 358, in run_mqtt_client
> mqtt_client.loop_forever()
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1756, in loop_forever
> rc = self._loop(timeout)
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1164, in _loop
> rc = self.loop_read()
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1556, in loop_read
> rc = self._packet_read()
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 2439, in _packet_read
> rc = self._packet_handle()
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3033, in _packet_handle
> return self._handle_publish()
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3327, in _handle_publish
> self._handle_on_message(message)
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3570, in _handle_on_message
> on_message(self, self._userdata, message)
> File "/usr/src/app/./index.py", line 215, in on_message
> snapshot_url = f"{frigate_url}/api/events/{frigate_event}/snapshot.jpg"
> UnboundLocalError: local variable 'frigate_event' referenced before assignment
> 2024-01-14 11:55:40,242 - __main__ - INFO - Time: 2024-01-14 11:55:40.242
> 2024-01-14 11:55:40,242 - __main__ - INFO - Python Version: 3.9.18 (main, Jan 11 2024, 11:28:41)
> 
> [GCC 12.2.0]
> 2024-01-14 11:55:40,242 - __main__ - INFO - Frigate Plate Recognizer Version: 1.8.3
> 
> 2024-01-14 11:55:40,242 - __main__ - INFO - Using CodeProject.AI API
> 2024-01-14 11:55:40,242 - __main__ - INFO - Starting MQTT client. Connecting to: 192.168.2.20
> 2024-01-14 11:55:40,243 - __main__ - INFO - MQTT Connected
> ```
> 
> Nothing is being saved into the plates folder

Yep I’m getting the same error. I think it’s because @ljmerza moved quite a bit of code around in the final update.

**Comment by gadget-man** [comment]
> @gadget-man how are the plates being stored? are they in folders or just all dumped into the same folder?
> 
> I get roughly about 400 per day. So it would end up being a very stacked folder if they are. My log is 7600 lines after 11 days...
> 
> If there is nothing in place currently, maybe a folder every month ?

Currently, it just saves a timestamp file into a single folder. The ability to include the numberplates in the file. Name has also been taken out in 1.8.3. Perhaps it makes sense to add an optional configuration flag, which would create sub folders for individual plates and/or months. If you’re getting a lot of hits, I would recommend setting up a separate script to archive/delete all the files on a periodic basis using crontab. I don’t think this is something that would be best placed being managed within the script itself.

**Comment by kyle4269** [comment]
> > Thanks for looking at it @ljmerza and doing the new builds. I am testing it now...
> 
> > 
> 
> > I am getting the following error continually every time a plate is detected:
> 
> > 
> 
> > ```
> 
> > 2024-01-14 11:54:51,040 - __main__ - INFO - Time: 2024-01-14 11:54:51.040
> 
> > 2024-01-14 11:54:51,040 - __main__ - INFO - Python Version: 3.9.18 (main, Jan 11 2024, 11:28:41)
> 
> > 
> 
> > [GCC 12.2.0]
> 
> > 2024-01-14 11:54:51,040 - __main__ - INFO - Frigate Plate Recognizer Version: 1.8.3
> 
> > 
> 
> > 2024-01-14 11:54:51,040 - __main__ - INFO - Using CodeProject.AI API
> 
> > 2024-01-14 11:54:51,040 - __main__ - INFO - Starting MQTT client. Connecting to: 192.168.2.20
> 
> > 2024-01-14 11:54:51,041 - __main__ - INFO - MQTT Connected
> 
> > Traceback (most recent call last):
> 
> > File "/usr/src/app/./index.py", line 403, in <module>
> 
> > main()
> 
> > File "/usr/src/app/./index.py", line 399, in main
> 
> > run_mqtt_client()
> 
> > File "/usr/src/app/./index.py", line 358, in run_mqtt_client
> 
> > mqtt_client.loop_forever()
> 
> > File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1756, in loop_forever
> 
> > rc = self._loop(timeout)
> 
> > File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1164, in _loop
> 
> > rc = self.loop_read()
> 
> > File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1556, in loop_read
> 
> > rc = self._packet_read()
> 
> > File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 2439, in _packet_read
> 
> > rc = self._packet_handle()
> 
> > File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3033, in _packet_handle
> 
> > return self._handle_publish()
> 
> > File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3327, in _handle_publish
> 
> > self._handle_on_message(message)
> 
> > File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3570, in _handle_on_message
> 
> > on_message(self, self._userdata, message)
> 
> > File "/usr/src/app/./index.py", line 215, in on_message
> 
> > snapshot_url = f"{frigate_url}/api/events/{frigate_event}/snapshot.jpg"
> 
> > UnboundLocalError: local variable 'frigate_event' referenced before assignment
> 
> > 2024-01-14 11:55:40,242 - __main__ - INFO - Time: 2024-01-14 11:55:40.242
> 
> > 2024-01-14 11:55:40,242 - __main__ - INFO - Python Version: 3.9.18 (main, Jan 11 2024, 11:28:41)
> 
> > 
> 
> > [GCC 12.2.0]
> 
> > 2024-01-14 11:55:40,242 - __main__ - INFO - Frigate Plate Recognizer Version: 1.8.3
> 
> > 
> 
> > 2024-01-14 11:55:40,242 - __main__ - INFO - Using CodeProject.AI API
> 
> > 2024-01-14 11:55:40,242 - __main__ - INFO - Starting MQTT client. Connecting to: 192.168.2.20
> 
> > 2024-01-14 11:55:40,243 - __main__ - INFO - MQTT Connected
> 
> > ```
> 
> > 
> 
> > Nothing is being saved into the plates folder
> 
> 
> 
> Yep I’m getting the same error. I think it’s because @ljmerza moved quite a bit of code around in the final update. 

frigate_event on 239 needs to be moved up before 215. I'm unable to test right now though.

**Comment by DrSpaldo** [comment]
> Yep I’m getting the same error. I think it’s because @ljmerza moved quite a bit of code around in the final update.

@gadget-man was it working in any of the other versions for you? ie 1.8.2?

> 
> Currently, it just saves a timestamp file into a single folder. The ability to include the numberplates in the file. Name has also been taken out in 1.8.3. Perhaps it makes sense to add an optional configuration flag, which would create sub folders for individual plates and/or months. If you’re getting a lot of hits, I would recommend setting up a separate script to archive/delete all the files on a periodic basis using crontab. I don’t think this is something that would be best placed being managed within the script itself.

No worries, no need to make it too complicated then, I can setup a script.

@ljmerza would it be possible to retain the feature @gadget-man added and keep the car registration plate in the file name

**Comment by gadget-man** [comment]
No I've not managed to get any of the 1.8.x versions working. I've reverted to my forked Repo and am testing a further update.

Once we've got the initial version stabilised using @ljmerza's more eloquent code, I may do a further PR for some enhancements.

**Comment by ljmerza** [comment]
been writing a ton of tests to avoid this in the future. pushed the latest with those tests and changes

**Comment by DrSpaldo** [comment]
> been writing a ton of tests to avoid this in the future. pushed the latest with those tests and changes

Still getting errors in 1.8.4

```
2024-01-15 12:10:55,710 - __main__ - INFO - Python Version: 3.9.18 (main, Jan 11 2024, 11:28:41)

[GCC 12.2.0]
2024-01-15 12:10:55,710 - __main__ - INFO - Frigate Plate Recognizer Version: 1.8.4

2024-01-15 12:10:55,711 - __main__ - INFO - Using CodeProject.AI API
2024-01-15 12:10:55,711 - __main__ - INFO - Starting MQTT client. Connecting to: 192.168.2.20
2024-01-15 12:10:55,711 - __main__ - INFO - MQTT Connected
Traceback (most recent call last):
File "/usr/src/app/./index.py", line 433, in <module>
main()
File "/usr/src/app/./index.py", line 429, in main
run_mqtt_client()
File "/usr/src/app/./index.py", line 388, in run_mqtt_client
mqtt_client.loop_forever()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1756, in loop_forever
rc = self._loop(timeout)
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1164, in _loop
rc = self.loop_read()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1556, in loop_read
rc = self._packet_read()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 2439, in _packet_read
rc = self._packet_handle()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3033, in _packet_handle
return self._handle_publish()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3327, in _handle_publish
self._handle_on_message(message)
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3570, in _handle_on_message
on_message(self, self._userdata, message)
File "/usr/src/app/./index.py", line 259, in on_message
if check_first_message(message):
TypeError: check_first_message() takes 0 positional arguments but 1 was given
2024-01-15 12:11:07,839 - __main__ - INFO - Time: 2024-01-15 12:11:07.839
```

**Comment by DrSpaldo** [comment]
@ljmerza still getting errors in 1.8.7

```
2024-01-15 16:04:55,493 - __main__ - INFO - Python Version: 3.9.18 (main, Jan 11 2024, 11:28:41)

[GCC 12.2.0]
2024-01-15 16:04:55,493 - __main__ - INFO - Frigate Plate Recognizer Version: 1.8.7

2024-01-15 16:04:55,493 - __main__ - INFO - Using CodeProject.AI API
2024-01-15 16:04:55,493 - __main__ - INFO - Starting MQTT client. Connecting to: 192.168.2.20
2024-01-15 16:04:55,494 - __main__ - INFO - MQTT Connected
Traceback (most recent call last):
File "/usr/src/app/./index.py", line 443, in <module>
main()
File "/usr/src/app/./index.py", line 439, in main
run_mqtt_client()
File "/usr/src/app/./index.py", line 398, in run_mqtt_client
mqtt_client.loop_forever()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1756, in loop_forever
rc = self._loop(timeout)
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1164, in _loop
rc = self.loop_read()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1556, in loop_read
rc = self._packet_read()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 2439, in _packet_read
rc = self._packet_handle()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3033, in _packet_handle
return self._handle_publish()
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3327, in _handle_publish
self._handle_on_message(message)
File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3570, in _handle_on_message
on_message(self, self._userdata, message)
File "/usr/src/app/./index.py", line 341, in on_message
plate_number, plate_score = get_plate(snapshot, after_data, license_plate_attribute)
NameError: name 'license_plate_attribute' is not defined
```

**Comment by kyle4269** [comment]
> @ljmerza still getting errors in 1.8.7
> 
> 
> 
> ```
> 
> 2024-01-15 16:04:55,493 - __main__ - INFO - Python Version: 3.9.18 (main, Jan 11 2024, 11:28:41)
> 
> 
> 
> [GCC 12.2.0]
> 
> 2024-01-15 16:04:55,493 - __main__ - INFO - Frigate Plate Recognizer Version: 1.8.7
> 
> 
> 
> 2024-01-15 16:04:55,493 - __main__ - INFO - Using CodeProject.AI API
> 
> 2024-01-15 16:04:55,493 - __main__ - INFO - Starting MQTT client. Connecting to: 192.168.2.20
> 
> 2024-01-15 16:04:55,494 - __main__ - INFO - MQTT Connected
> 
> Traceback (most recent call last):
> 
> File "/usr/src/app/./index.py", line 443, in <module>
> 
> main()
> 
> File "/usr/src/app/./index.py", line 439, in main
> 
> run_mqtt_client()
> 
> File "/usr/src/app/./index.py", line 398, in run_mqtt_client
> 
> mqtt_client.loop_forever()
> 
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1756, in loop_forever
> 
> rc = self._loop(timeout)
> 
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1164, in _loop
> 
> rc = self.loop_read()
> 
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 1556, in loop_read
> 
> rc = self._packet_read()
> 
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 2439, in _packet_read
> 
> rc = self._packet_handle()
> 
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3033, in _packet_handle
> 
> return self._handle_publish()
> 
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3327, in _handle_publish
> 
> self._handle_on_message(message)
> 
> File "/usr/local/lib/python3.9/site-packages/paho/mqtt/client.py", line 3570, in _handle_on_message
> 
> on_message(self, self._userdata, message)
> 
> File "/usr/src/app/./index.py", line 341, in on_message
> 
> plate_number, plate_score = get_plate(snapshot, after_data, license_plate_attribute)
> 
> NameError: name 'license_plate_attribute' is not defined
> 
> ```

I put in a PR to fix this. I ran a few tests on my camera before the PR and everything seemed to work. Should be good once 1.8.8 is approved.

**Comment by kyle4269** [comment]
@DrSpaldo 1.8.8 was just released.

**Comment by DrSpaldo** [comment]
> @DrSpaldo 1.8.8 was just released.

Thanks @kyle4269 . Good news is that I am not getting the errors anymore and it is logging some of the plates.

Bad news is that none of the snapshots are being saved into the folder :(

**Comment by kyle4269** [comment]
> > @DrSpaldo 1.8.8 was just released.
> 
> Thanks @kyle4269 . Good news is that I am not getting the errors anymore and it is logging some of the plates.
> 
> Bad news is that none of the snapshots are being saved into the folder :(

What's your docker-compose look like?

**Comment by DrSpaldo** [comment]
> 
> What's your docker-compose look like? This is what I have
> 
> ```
> volumes:
>   - /cpai/dev/config/plates:/plates:rw
> ```
> 
> try:
> 
> ```
>  - /mnt/user/appdata/plate-recognizer/plates:/plates:rw
> ```

That got it working. Thanks @kyle4269 

Interesting that the plates folder didn't need the preceding /usr/src/app/ like the config folder did.

For anyone else that encounters the issue that is using Unraid in a stack, here is my config:

```
services:
  frigate_plate_recognizer:
    image: lmerza/frigate_plate_recognizer:latest
    container_name: frigate_plate_recognizer
    volumes:
      - /mnt/user/appdata/plate-recognizer:/usr/src/app/config
      - /mnt/user/appdata/plate-recognizer/plates:/usr/src/app/plates:rw
      - /mnt/user/appdata/plate-recognizer/plates:/plates:rw
    restart: unless-stopped
    environment:
      - TZ=Australia/Brisbane
```

There is probably no need to have the plates mapped to two locations, however, I just did it incase.

**Comment by kyle4269** [comment]
> > 
> 
> > What's your docker-compose look like? This is what I have
> 
> > 
> 
> > ```
> 
> > volumes:
> 
> >   - /cpai/dev/config/plates:/plates:rw
> 
> > ```
> 
> > 
> 
> > try:
> 
> > 
> 
> > ```
> 
> >  - /mnt/user/appdata/plate-recognizer/plates:/plates:rw
> 
> > ```
> 
> 
> 
> That got it working. Thanks @kyle4269 
> 
> 
> 
> Interesting that the plates folder didn't need the preceding /usr/src/app/ like the config folder did.
> 
> 
> 
> For anyone else that encounters the issue that is using Unraid in a stack, here is my config:
> 
> 
> 
> ```
> 
> services:
> 
>   frigate_plate_recognizer:
> 
>     image: lmerza/frigate_plate_recognizer:latest
> 
>     container_name: frigate_plate_recognizer
> 
>     volumes:
> 
>       - /mnt/user/appdata/plate-recognizer:/usr/src/app/config
> 
>       - /mnt/user/appdata/plate-recognizer/plates:/usr/src/app/plates:rw
> 
>       - /mnt/user/appdata/plate-recognizer/plates:/plates:rw
> 
>     restart: unless-stopped
> 
>     environment:
> 
>       - TZ=Australia/Brisbane
> 
> ```
> 
> 
> 
> There is probably no need to have the plates mapped to two locations, however, I just did it incase.

Yeah, I couldn't get the plates to save either.. So I changed a piece of code to look at that location. Glad it's working for you!

**Comment by DrSpaldo** [comment]
For reference, I wanted to have the files moved into a monthly folder, as I will get 200+ per day. So a friend helped me write a script, see below if anyone else wants to use it...

```
#!/usr/bin/env bash

# Comment out to actually do stuff
#DUMMY=echo

set -e

if [ -z "$1" ]; then
	echo "Must pass directory name to process as first argument"
	exit 1
fi
if [ -z "$2" ]; then
	echo "Must pass destination directory name as second argument"
	exit 1
fi

find "$1" -maxdepth 1 -type f -iname \*.png | ( while [ 1 ]; do
	read fname
	if [ $? -ne 0 ]; then
		break
	fi
	if [ -z "$fname" ]; then
		continue
	fi
	# Get modification time of file as epoch timestamp
	tstamp=$(stat -c %Y "$fname")
	# Create destination directory name using timestamp
	dirname="$2/$(date -d @$tstamp +%Y/%m)"
	if [ ! -e "$dirname" ]; then
		$DUMMY mkdir -p "$dirname"
		if [ $? -ne 0 ]; then
			echo "Unable to create $dirname, aborting"
			exit 1
		fi
	fi
	$DUMMY mv "$fname" "$dirname"
	if [ $? -ne 0 ]; then
		echo "Unable to move $fname to $dirname, aborting"
		exit 1
	fi
done )
```

**Comment by gadget-man** [comment]
Same issue here on the config. On a synology system, it now seems to be inconsistent between how the config folder is mapped in docker-compose.yml and how you need to map the plates directory, which is somewhat confusing.

I think the issue is because in index.py, the `CONFIG_PATH` is a relative path i.e.: `./config/config.yml`, whereas `SNAPSHOT_PATH` has been changed to absolute: `/plates`

**Comment by gadget-man** [comment]
I've also noticed that neither `draw_box` nor `always_save_snapshot` are working as expected. For draw box, I think it's because the code now only saves a snapshot of the cropped image (rather than the full snapshot), so the box is probably being drawn outside the frame of the snapshot.  I can't easily see why `always_save_snapshot` isn't working as expected.

**Comment by kyle4269** [comment]
> I've also noticed that neither `draw_box` nor `always_save_snapshot` are working as expected. For draw box, I think it's because the code now only saves a snapshot of the cropped image (rather than the full snapshot), so the box is probably being drawn outside the frame of the snapshot.  I can't easily see why `always_save_snapshot` isn't working as expected.
> 
> 
> 
> 

I'm pretty sure if line 222 crop is set to 0 it'll no longer crop the image. 

I've ran into the same issue with the box being drawn outside the frame of the snapshot. If you compare the snapshot that's saved to the one frigate has under events.  You'll see they're different. 

I've talked to Nick about this a few times. He suggested a couple things. Wait about 5 seconds for the event to end, then saving the snapshot. I currently do this and my snapshot will be the exact same as frigate. But the box is still drawn outside the frame. 

Watch for the events "end_time" 
and save the snapshot then. 

For me, I'm just grabbing plates from my driveway, so waiting 5 seconds works for me. BUT those who are getting plates from the road, that probably wouldn't be a good route. 

So what I ended up doing was waiting 5 seconds and editing the image with the plate number under the date and time. It then sends it to my telegram.

**Comment by gadget-man** [comment]
I think the issue is that frigate returns the dimensions of the plate based on the full sized snapshot, which we are then applying to the bottom corner of the cropped image. I agree, setting crop=0 should solve this, but not sure how much that affects the efficiency of plate_recognizer?  I think this is compounded by the fact that frigate may update the snapshot image for the event several times, so they are slightly out of sync - not sure there's anything that can be done about that.

I've also started noticing a new issue that I hadn't come across - I'm frequently seeing a `duplicated snapshot from Frigate as top_score from before and after are the same` appearing several times, including the very first mqtt entry, which results in the entire event being missed from detection. Anyone else seeing the same?  I don't recall coming across this in v 1.7.X versions.

**Comment by gadget-man** [comment]
> could we retrieve the image again but with crop=0 just for the image save? I cant easily test this since cars dont pass by my cameras too often hence the need for the automated tests

That's what I did in the original PR - for save_image it retrieved a new copy of the snapshot and applied the box to that. However, I don't think that would solve the issue of the 'box' dimensions in the matt being different from the snapshot, because there could be a few 100ms delay between the script processing the mqtt message and requesting the snapshot from frigate, by which time Frigate may have updated the snapshot (and a later mqtt message would therefore have more updated box coordinates).  I wonder if there's a way to interrogate Frigate to get the box dimensions for an specific event - in which case you could do it based on the delay concept (or wait until the event has an end time)

**Comment by ljmerza** [comment]
Yes the delay workaround was brought up before but wasn't a very clean fix since if any event lasted longer it would not work. Ideally there's some indication when an event has ended and we'd use that to wait to get the image

**Comment by gadget-man** [comment]
I think we might be able to overcome this by calling the api/events call to retrieve updated data, rather than using the data from the `after` payload. I'll have a play with this. If so, it would make sense to move the `save_image` function back to the end of the script (after posting to mqtt) so that this provess doesn't delay the overall mqtt updates.

**Comment by gadget-man** [comment]
> Once "end_time" is populated with a time. The event is officially over and stored into frigates db. Any thoughts on that?

In the vast majority of situations I think this would work, however if a vehicle arrives in frame and then parks, the event would stay 'live'. What we really want is just to grab a snapshot and it's associated event details at the same time. I think the best way to do this is to call a new response.get for a full snapshot and then an api call to get the event details immediately afterwards (or before) and then use that data. It is likely that the snapshot would be a slightly different Frame from that used by plate_recognizer to detect the plate, but at least the box would align....

**Comment by kyle4269** [comment]
Here is an example. 

Waiting 5 seconds:
![image](https://github.com/ljmerza/frigate_plate_recognizer/assets/18011470/46bdfde6-422f-4208-8b43-721d2cb3ae5b)

Not waiting:
![image](https://github.com/ljmerza/frigate_plate_recognizer/assets/18011470/bf5f915f-a905-4a34-9984-4a34569433cd)

**Comment by kyle4269** [comment]
> Yes the delay workaround was brought up before but wasn't a very clean fix since if any event lasted longer it would not work. Ideally there's some indication when an event has ended and we'd use that to wait to get the image

I was driving up and down my driveway testing it myself.. Until I found out you can manually run clips through Frigate..  I copied all the settings from my LPR config to my "test" see below.. Makes testing so much easier.

https://docs.frigate.video/development/contributing/#2-modify-your-local-config-file-for-testing

```
cameras:
   test:
     ffmpeg:
        inputs:
          - path: /media/frigate/testing/TundraFrontDay.mp4
            input_args: -re -stream_loop -1 -fflags +genpts
```

**Comment by gadget-man** [comment]
I'm looking at the alternative strategy where we grab a new (uncrossed) snapshot, as well as the event data using `GET /api/events/<id>`. However, I've noticed that unlike the mqtt event data, the box coordinates returned in the API call are all decimals (e.g. `[0.28828125, 0.42916666666666664, 0.14140625, 0.08194444444444444]`). I can't find a way to translate those co-ordinates into sensible numbers relative to the image size to draw a box. I've compared them against the currently returned coordinates that we're using to draw the box in save_image and can't see any correlation.  Anybody know how we turn these co-ordinates into something appropriate to then draw a box using them?

**Comment by kyle4269** [comment]
@gadget-man 
Looks at the way Nick does it https://github.com/blakeblackshear/frigate/discussions/8165 It looks like he gets the even data the same way you're trying. He also does a calculation to find the coordinates for the image. I believe he's finding the coordinates of the license plate then sending that cropped image to CPAI. Maybe this will help?
```
# find coordinates of image to send to detector
x_mid = (box[0] + (box[2] / 2)) * known_res[0]
y_max = (box[1] + box[3]) * known_res[1]

cropped_image = img.crop((max(0, x_mid - 150), 0, min(known_res[0], x_mid + 150), known_res[1]))
```

**Comment by DrSpaldo** [comment]
Can there be a consideration to appending the registration to the end of the file? Or if it’s not too hard having option/s?

**Comment by ljmerza** [comment]
I was hesitant to go down the rabbit hole of configuring the file name so kept it static.

**Comment by kyle4269** [comment]
@gadget-man 
After thinking about this for a while.. I looked at the response data from CP.AI and below is what I have.

```
{
   "confidence":0.8886001706123352,
   "label":"Plate: 4260",
   "plate":"4260",
   "x_min":1217,
   "y_min":501,
   "x_max":1392,
   "y_max":618
}
```
CP.AI already has the coordinates of the plate. I changed your code to:
```
    vehicle = (
        x_min,
        y_min,
        x_max,
        y_max
    )
```
Here is the result:
![photo_4990253133661842857_w](https://github.com/ljmerza/frigate_plate_recognizer/assets/18011470/93c61ba3-ba7b-48fb-acea-26fd84a29323)

Plate Recognizer has a similar response.. https://guides.platerecognizer.com/docs/snapshot/api-reference/#read-number-plates-from-an-image

**Comment by DrSpaldo** [comment]
449 snapshots yesterday!  Might need to come up with a better filing system than per month.

Many are cropped and are missing the plate, would be great to get the full shot...

**Comment by kyle4269** [comment]
> 449 snapshots yesterday! Might need to come up with a better filing system than per month.
> 
> Many are cropped and are missing the plate, would be great to get the full shot...

Wow, you're getting a good amount of plates! I put in a PR for this. You'll need to add "crop_image: false" to your config under frigate once this is approved.

**Comment by gadget-man** [comment]
I've been working on this, and by repeating the call for a snapshot and then immediately using the API to get the event data, the box dimensions are MUCH more accurate. I've also changed it so that it saves a full snapshot, not just the crop. Will push a PR for this shortly. @kyle4269 I think this would negate the need to have a separate flag for crop_image - because the box dimensions published by frigate are for the full snapshot (not the crop), we're going back to saving the full snapshot.

**Comment by kyle4269** [comment]
> I've been working on this, and by repeating the call for a snapshot and then immediately using the API to get the event data, the box dimensions are MUCH more accurate. I've also changed it so that it saves a full snapshot, not just the crop. Will push a PR for this shortly. @kyle4269 I think this would negate the need to have a separate flag for crop_image - because the box dimensions published by frigate are for the full snapshot (not the crop), we're going back to saving the full snapshot.

Heck of a job man!! I'm gonna try to find some time this weekend to test.

**Comment by kyle4269** [comment]
@gadget-man Did a little testing with your 1.8.9 build. See the below image, It's a good distance off from the actual plate.

![4269_test_2024-01-19_12-15-15](https://github.com/ljmerza/frigate_plate_recognizer/assets/18011470/5deebf9e-c4f9-4004-9d10-a9bd5f711448)

**Comment by gadget-man** [comment]
Yeah, I was getting an almost perfect match. Maybe I was just lucky! From the logs, can you see time time difference between getting the updated snapshot and drawing the box?

**Comment by gadget-man** [comment]
I've had another thought - plate_recognizer also returns box dimensions (not sure if CodeProject.ai also does the same?). Perhaps we can use those to draw the box, rather than the data from Frigate? The issue would be the crop - unless we can reliably get the dimensions used to create the crop, we'd need to send the full image to platerecognizer to get consistent co-ordinates back....

**Comment by gadget-man** [comment]
> See my post [#21 (comment)](https://github.com/ljmerza/frigate_plate_recognizer/issues/21#issuecomment-1897644310) about using that. I think if we use CPAI and Plate Recognizer coordinates after the image was processed, we'd then have exact coordinates for the plate.

Sorry I'd totally missed that! Had you changed the code to send the un-cropped image for recognition in your updates?

**Comment by kyle4269** [comment]
All good man! Yeah, I don't send the images cropped. But thinking about it.. It shouldn't matter if we send cropped images or not. If we aren't using Frigate coordinates. We send the image to be processed and receive the coordinates for either CPAI or Plate Recognizer. Those should be exact to whatever we send. I think?

**Comment by gadget-man** [comment]
Yes, but if you send a cropped image for processing, if you choose to save the snapshot that would have to be cropped as well. I'm not sure why the original code only sends the cropped image - presumably for speed of processing / accuracy improvements? Personally, I'd prefer to have the final snapshot that is saved as the uncropped version.

**Comment by kyle4269** [comment]
That's what I figured cropping it was for. I do like having the uncropped version also. What about if the snapshot is saved then sent, instead of sending it directly from Frigate? In theory we'd be 100% accurate with the coordinates on every image. Wouldn't matter them if its cropped or not.. My only concern would on a busy street, would Plate Recognizer be able to keep up?

**Comment by kyle4269** [comment]
I turned on Plate Recognizer to get the response data.
```
{
   "processing_time":52.266,
   "results":[
      {
         "box":{
            "xmin":1454,
            "ymin":277,
            "xmax":1553,
            "ymax":322
         },
         "plate":"4269",
         "region":{
            "code":"us-ri",
            "score":0.103
         },
         "score":0.9,
         "candidates":[
            {
               "score":0.9,
               "plate":"4269"
            }
         ],
         "dscore":0.86,
         "vehicle":{
            "score":0.392,
            "type":"Big Truck",
            "box":{
               "xmin":1182,
               "ymin":1,
               "xmax":1912,
               "ymax":431
            }
         }
      }
   ],
   "filename":"1805_APZEs_upload.jpg",
   "version":1,
   "camera_id":"None",
   "timestamp":"2024-01-19T18:05:14.280348Z"
}
```

**Comment by gadget-man** [comment]
Ok I've got some feedback from Nick over at Frigate. We might be better using latest.jpg for the identified camera, rather than snapshot.jpg.

More work required!  I'll try and find time to have another play at this over the weekend.

**Comment by gadget-man** [comment]
Based on the feedback back from Nick, I think the best approach would be as follows:

1. On mqtt message being received, grab the latest image using `GET /api/<camera_name>/latest.jpg` (note this returns a full size image, no option to get it cropped).
2. Based on the box dimensions of the object (or plate) create our own crop and pass this to plate_recognizer / CP.AI.
3. If a plate is returned, save the original image with the box dimensions of the plate from plate_recognizer / CP.AI overlaid. Note because we've created our own crop and only have 1 original image from Frigate, we can work out how to transpose the coordinates from the cropped image back onto the full size image and save it.

Bit of re-working the current code, but sounds like it might be the best solution. As a starting point, I'd probably try and run a parallel routine to just grab `latest.jpg` and save the Frigate 'box' coordinates onto it, just to see how close they are.

Any thoughts?

**Comment by gadget-man** [comment]
@kyle4269  looking at my snapshots from today, I've not had very many, but every one has been an EXACT match between the red box we've drawn and the original plate. Can I just check: are you using a test script/video for the examples above, and if so are you sure that the resolution of the saved clip matches the resolution of the 'detect' camera in your Frigate configuration? I had an issue the other day where I downloaded a clip to use for testing, forgetting that my clips are saved at a higher resolution than the detect stream, resulting in all my boxes being out of alignment....

**Comment by kyle4269** [comment]
> @kyle4269  looking at my snapshots from today, I've not had very many, but every one has been an EXACT match between the red box we've drawn and the original plate. Can I just check: are you using a test script/video for the examples above, and if so are you sure that the resolution of the saved clip matches the resolution of the 'detect' camera in your Frigate configuration? I had an issue the other day where I downloaded a clip to use for testing, forgetting that my clips are saved at a higher resolution than the detect stream, resulting in all my boxes being out of alignment....

I am using a clone from your GitHub for testing. The clip im using is a frigate event clip of a vehicle coming into the driveway. Also for the "test" camera config I used was exactly what I have for my LPR. So I guess it's possible that when downloading the clip from frigate it plays with the resolution.

**Comment by gadget-man** [comment]
OK how strange. And your `-detect` and `-record` streams in your Frigate config.yml are the same stream?  It would be very interesting to know if you have the same issue on your 'real' camera.

**Comment by kyle4269** [comment]
> OK how strange. And your `-detect` and `-record` streams in your Frigate config.yml are the same stream?  It would be very interesting to know if you have the same issue on your 'real' camera.

That's correct. My record and detect are both on my main stream. I will try it on my real camera tonight. Have to wait for the kid to go to bed because I need to go drive up and down a few time haha

**Comment by kyle4269** [comment]
@gadget-man 
Here is an example from my real camera 
![LPR_2024-01-19_17-05-36](https://github.com/ljmerza/frigate_plate_recognizer/assets/18011470/59cb62d4-fe46-45cf-8abf-41cf5969af4c)

**Comment by gadget-man** [comment]
I’ve had a couple of detections today and they are a perfect match. Yours always seem to be offset by a very similar amount which is also a bit strange. I’m still suspicious that it’s a resolution issue. What resolution have you specified for Detect in your Frigate config.yml?
Regards,
Paul Needler
CEO
[Logo]<https://www.iparcelbox.com/>
NOTE OUR ADDRESS FOR ALL DELIVERIES AND CORRESPONDENCE HAS CHANGED:
Pine Trees, 9A Westfield Park, Elloughton, E. Yorkshire, HU15 1AN, UK
Tel: +44 (0)7539 465747
Web: www.iParcelBox.com<http://www.iparcelbox.com/>
iParcelBox Ltd | Registered in England and Wales
Company No. 11622993 | VAT No. GB314127736

[Facebook icon]<https://www.facebook.com/iparcelbox>  [Instagram icon] <https://www.instagram.com/iparcelbox>   [Twitter icon] <https://twitter.com/iparcelbox>   [Twitter icon] <https://linkedin.com/company/iparcelbox>
[Made in Britain]<https://www.madeinbritain.org/members/iparcelbox-ltd>
[HullLive Awards Winner]
[Trustpilot rating]
[Trustpilot Stars]<https://uk.trustpilot.com/review/iparcelbox.com?utm_medium=Trustbox&utm_source=EmailSignature2>
[Trustpilot Logo]<https://uk.trustpilot.com/review/iparcelbox.com?utm_medium=Trustbox&utm_source=EmailSignature2>
This communication may contain privileged or confidential information which is for the exclusive use of the intended recipient. If you are not the intended recipient, please note that you may not distribute or use this communication or the information it contains. If this email has reached you in error, please delete it and any attachment. Internet communications are not secure and iParcelBox Ltd does not accept legal responsibility for the content of this message.


On 19 Jan 2024, at 22:37, kyle4269 ***@***.***> wrote:

﻿

@gadget-man<https://github.com/gadget-man>
Here is an example from my real camera
LPR_2024-01-19_17-05-36.png (view on web)<https://github.com/ljmerza/frigate_plate_recognizer/assets/18011470/59cb62d4-fe46-45cf-8abf-41cf5969af4c>

**Comment by kyle4269** [comment]
Running at 1920x1080. It's very odd. If everyone else is getting perfect matches, it's a me issue haha. Im perfectly fine changing the code on new releases to grab the coordinates from CPAI. So don't keep banging away at the keyboard for me. I appreciate all the work you are putting in. Wish I was as knowledgeable as you in python.  Learning a lot from you and Leonardo.
![image](https://github.com/ljmerza/frigate_plate_recognizer/assets/18011470/03789def-9caf-4cf2-b245-9cdbe408faf1)

**Comment by kyle4269** [comment]
```
  LPR:
    ffmpeg:
      inputs:
      - path: rtsp://554/cam/realmonitor?channel=1&subtype=1
        roles:
        - audio
      - path: rtsp://554/cam/realmonitor?channel=1&subtype=0
        roles:
        - record
        - detect
      output_args:
        record: preset-record-generic-audio-aac
    detect:
      width: 1920
      height: 1080
      fps: 15
      stationary:
        interval: 10
    objects:
      track:
      - person
      - car
      - face
      - license_plate
         # - amazon
         # - ups
         # - fedex
      filters:
       car:
         min_score: .50
         threshold: .80
       license_plate:
         min_score: .6
    motion:
      threshold: 85
      improve_contrast: true
      contour_area: 30
    zones:
      Lp:
        coordinates: 0,382,0,1080,1920,1080,1920,400
        objects:
        - car
        - person
        - face
        - license_plate
    snapshots:
      enabled: true
      clean_copy: true
      bounding_box: false
      crop: false
      required_zones:
      - Lp
    mqtt:
      bounding_box: false
      crop: false
      height: 1080
      quality: 100
    record:
      enabled: true
      events:
        required_zones:
        - Lp
```

**Comment by gadget-man** [comment]
Doesn’t really make sense does it?  In my latest code, if you add a line to log image_width and image_height, does it correctly return 1920 and 1080?  And if you interrogate the saved snapshot jpg file, has it also saved at those dimensions?

On 20 Jan 2024, at 13:40, kyle4269 ***@***.***> wrote:

﻿

  LPR:
    ffmpeg:
      inputs:
      - path: rtsp://554/cam/realmonitor?channel=1&subtype=1
        roles:
        - audio
      - path: rtsp://554/cam/realmonitor?channel=1&subtype=0
        roles:
        - record
        - detect
      output_args:
        record: preset-record-generic-audio-aac
    detect:
      width: 1920
      height: 1080
      fps: 15
      stationary:
        interval: 10
    objects:
      track:
      - person
      - car
      - face
      - license_plate
         # - amazon
         # - ups
         # - fedex
      filters:
       car:
         min_score: .50
         threshold: .80
       license_plate:
         min_score: .6
    motion:
      threshold: 85
      improve_contrast: true
      contour_area: 30
    zones:
      Lp:
        coordinates: 0,382,0,1080,1920,1080,1920,400
        objects:
        - car
        - person
        - face
        - license_plate
    snapshots:
      enabled: true
      clean_copy: true
      bounding_box: false
      crop: false
      required_zones:
      - Lp
    mqtt:
      bounding_box: false
      crop: false
      height: 1080
      quality: 100
    record:
      enabled: true
      events:
        required_zones:
        - Lp

**Comment by gadget-man** [comment]
Could you DM me your test .mp4 file so that I can play with it here?  I’ll set up a test camera using your settings and see if I can work it out!

**Comment by kyle4269** [comment]
Doesn't make sense at all. It does say 1920x1080 for my snapshot. I'll put it on my google drive and send you a link when I get home. Where do you want me to dm it?

**Comment by gadget-man** [comment]
If you're happy to share it, just past the public google drive link here. Otherwise, you can email me at sales at iparcelbox dot com

**Comment by gadget-man** [comment]
Hi @kyle4269, so I just tried using the mp4 file you emailed me and the Frigate config you posted above (the only change being that I changed the input to the following for test purposes:
`
inputs:
       - path: /config/TundraFrontDay.mp4
         input_args: -re -stream_loop -1 -fflags +genpts
`
The plate was detected successfully, and this is the file that was saved by frigate_plate_recognizer:
![4269_LPR_2024-01-20_19-10-41](https://github.com/ljmerza/frigate_plate_recognizer/assets/24406912/9b3e9ae2-455b-4a32-a528-9f96235fff46)

I allowed the script to repeat quite a few times - each time the coordinates were very slightly different, but each time right smack bang in the middle of the plate.

Please can you check you're using the very latest version of the script from my forked repo, and try creating a test camera in Frigate replicating the above, to see if you can get similar results?

**Comment by kyle4269** [comment]
I'm using 0.13.0-rc1 and am using your latest fork. I will give it a try again and see my results. It looks like you only changed the stream loop to 1?

**Comment by gadget-man** [comment]
I'm on `0.13.0-c35c7da`. Let me also update to `rc1` just to make sure that doesn't make a difference...!

**Comment by kyle4269** [comment]
@gadget-man 
Can confirm that it's working!

**Comment by gadget-man** [comment]
Excellent!  Did you have to change any settings to get it working?  If so, perhaps there is something we could add to the readme to help others?

**Comment by kyle4269** [comment]
I didn't change anything other than in my testing camera 0 to -1.. But that doesn't matter for my main camera.

**Comment by gadget-man** [comment]
Ok great. Glad it's working at least!  Next mini-project is to try and implement some kind of 'watched plates' function.

**Comment by kyle4269** [comment]
> Ok great. Glad it's working at least!  Next mini-project is to try and implement some kind of 'watched plates' function.

I like that! Can have it send a mqtt message if the plate matches.  This was HA can grab that and send out notifications?

**Comment by gadget-man** [comment]
> > Ok great. Glad it's working at least!  Next mini-project is to try and implement some kind of 'watched plates' function.
> 
> I like that! Can have it send a mqtt message if the plate matches. This was HA can grab that and send out notifications?

I haven't really thought too much about it yet. I know Platerecognizer returns 'candidate' plates as well as the best match, not sure of CP.AI does the same. So we could test those against a list of watched plates. Or do some kind of fuzzy matching.   If a matched plate is found, it could either overwrite the recognised plate with the match, or perhaps add an additional tag to the existing MQTT message with the watched plate.

**Comment by kyle4269** [comment]
I looked into the API for CP.AI and I don't see any reference to possible matches.. I do have a clip of a plate that rarely comes back to the correct plate.. So when/if you get to that point of testing matching, I can give you that clip.

**Comment by gadget-man** [comment]
In 1.8.12, I'm finding that the red bounding box with the recognised plate is not longer appearing. I do see a small red dot in the top left corner. 

Looking at the code and your latest PR @ljmerza, I note that you converted the initial dimensions to int() before multiplying by the image width / height. I suspect this may be the cause - the returned value is expected to be between 0 and 1 so this may be the issue?

e.g.
`dimension_1 = int(final_attribute[0]['box'][0])` << this is almost certainly returning either 0 or 1.

I'm not at home to test it, but I think the int() on these 4 lines was perhaps added in error as part of your review of the PR?

**Comment by ljmerza** [comment]
ah youre right was trying to make sense of the math when making tests. fixed in 1.8.13
